### PR TITLE
CLOUD-3256 - Allow kubernetes to control probe retries; avoid probes taking longer than kubernetes timeout settings

### PR DESCRIPTION
**Pull request by luck3y**
https://issues.jboss.org/browse/CLOUD-3256
Signed-off-by: Ken Wills <kwills@redhat.com>

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [ ] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

File: templates/eap-cd-amq-persistent-s2i.json [hunk 1]
```diff
@@ -606,7 +606,8 @@
                                             "-c",
                                             "/opt/eap/bin/readinessProbe.sh"
                                         ]
-                                    }
+                                    },
+                                    "initialDelaySeconds": 10
```
Comment (bstansberry) [inline templates/eap-cd-amq-persistent-s2i.json:610]: Note: for the record. The old behavior baked into probe shell script would wait 5 secs before probing and then the probe would loop retrying with a 5 sec pause between attempts. In practice we don't observe EAP containers being ready in less than 10 secs anyway, so using 10 here avoids spurious initial failures without adding a pointless delay.

**Comment by bstansberry** [comment]
@luck3y Let's work through what the old probe behavior would have been (in 3.11) vs what it is now. Factoring in our template settings, k8s behavior, and our probe logic itself.

tl;dr There is a huge change in the time it takes to make a container as not live/ready. From roughly 8 mins to ~ 30 secs.

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes describes the k8s behavior; i.e. the config options and their default value.

https://docs.openshift.com/container-platform/3.11/dev_guide/application_health.html is more OpenShift specific. It's ambiguous about which k8s params work; i.e. saying in 1 place that only two work but then later showing a 4 in the example that actually uses the EAP probe.  It's clear though that timeoutSeconds does not work, and that is consistent with what we observed.

So the liveness probe BEFORE:

1) Wait 60 secs to start due to "initialDelaySeconds": 60
2) Execute probe
3) If exit code 0, default 'successThreshold' == 1 so k8s would consider the container live.
        A) Wait default 10 periodSeconds
        B) go to 2)
4) else, probe internally retries up to 30 times with a 5 sec sleep
    A) if anywhere in there it passes, probe exits with exit code 0
          i) Wait default 10 periodSeconds
         ii) go to 2)
    B) if it never passes, 30 * 5 secs have passed. exit code > 0
       i) Default failureThreshold is 3. So if this is the first or second consecutive failure, 
             a) Wait default 10 periodSeconds
             b) go to 2)
       ii) Else, container is marked as not live.  Min lapsed time since start 60 + 3 * ((30 * 5) + 10) = 540 secs. Min elapsed time since successful probe 3 * ((30 * 5) + 10) = 480 secs

The liveness probe NOW:

1) Wait 60 secs to start due to "initialDelaySeconds": 60
2) Execute probe
3) If exit code 0, default 'successThreshold' == 1 so k8s would consider the container live.
        A) Wait default 10 periodSeconds
        B) go to 2)
4) else, exit code > 0
    A) Default failureThreshold is 3. So if this is the first or second consecutive failure, 
             a) Wait default 10 periodSeconds
             b) go to 2)
    B) Else, container is marked as not live.  Min elapsed time since start 60 + 3 * 10 = 90 secs. Min elapsed time since successful probe 3 * 10 = 30 secs

The readiness probe BEFORE:

1) Execute probe
2) Waited 5 secs to start due to internal hard coded sleep
3) If exit code 0, default 'successThreshold' == 1 so k8s would consider the container ready.
        A) Wait default 10 periodSeconds
        B) go to 1)
4) else, probe internally retries up to 30 times with a 5 sec sleep
    A) if anywhere in there it passes, probe exits with exit code 0, 
          i) Wait default 10 periodSeconds
         ii) go to 1)
    B) if it never passes, 30 * 5 secs have passed. exit code > 0
       i) Default failureThreshold is 3. So if this is the first or second consecutive failure
             a) Wait default 10 periodSeconds
             b) go to 1)
       ii) Else, container is marked as not ready.  Elapsed time since start or a successful probe 3 * ((30 * 5) + 10) = 480 secs

The readiness probe NOW:

1) Wait 10 secs to start due to "initialDelaySeconds": 10
2) Execute probe
3) If exit code 0, default 'successThreshold' == 1 so k8s would consider the container ready.
        A) Wait default 10 periodSeconds
        B) go to 2)
4) else, exit code > 0
    A) Default failureThreshold is 3. So if this is the first or second consecutive failure, 
             a) Wait default 10 periodSeconds
             b) go to 2)
    B) Else, container is marked as not ready.  Min elapsed time since start 10 + 3 * 10 = 40 secs. Min elapsed time since successful probe 3 * 10 = 30 secs

**Comment by bstansberry** [comment]
My last comment for the BEFORE cases assumed the probe internally didn't result in Status.HARD_FAILURE, i.e. either Status.READY (the ok cases) or Status.FAILURE or Status.NOT_READY.

A HARD_FAILURE would omit the 30 * 5 secs internal probe retry loop. You'd still get the 3 failureThreshold retries with 10 periodSeconds pause in between.

So, why would you get a HARD_FAILURE?

os-eap-probes/2.0/added/probes/probe/eap/dmr.py

BootErrorsTest -- mgmt op returns success, result is not empty.
DeploymentTest -- mgmt op returns success, result includes FAILED for any deployment
HealthCheckTest -- 
    a) 2nd step in two step composite success but result is DOWN
    b) No 2nd step section in response, or its outcome != success, or it reports no result

Another thing that will break the loop is the HTTP request returning with a non 200 and the response not including a usable management response structure with a 'result' node list with at least as many elements as # of steps in the probe's composite op.  This is the path I'd expect to be followed while the server is starting; the request either won't connect or will go down the https://github.com/wildfly/wildfly-core/blob/master/controller/src/main/java/org/jboss/as/controller/ModelControllerImpl.java#L392 path

So, the cases where the probe executes correctly but the responses clearly indicate "bad", or cases where the probe fails with no useful data, expected during boot, results in no retries.

What leads to  Status.FAILURE?

ServerStatusTest -- mgmt op doesn't return success, has failure-description
BootErrorsTest -- mgmt op doesn't return success, has failure-description
DeploymentTest -- mgmt op doesn't return success, has failure-description OR result for a deployment != OK && != FAILED. So "NEW" (which never occurs) or "STOPPED" (for a disabled deployment)
HealthCheckTest -- mgmt op doesn't have a result or has no result for step-1  (read-resource to check for subsystem existence)

What leads to Status.NOT_READY?

ServerStatusTest -- mgmt op returns success, result != RUNNING. So STARTING, RELOAD_REQUIRED, RESTART_REQUIRED, STOPPING, STOPPED. For STARTING and STOPPED there shouldn't be a response, so really it's RELOAD_REQUIRED, RESTART_REQUIRED, STOPPING, STOPPED.

So for all those things the only one that indicates a state that there's any reason to expect a change if there's a retry is ServerStatusTest and NOT_READY.  And I don't see why retrying 30 times with a 5 sec delay is a reasonable way to handle that. The k8s 3 retries with the failureThreshold seems sufficient.

**Comment by bstansberry** [comment]
@luck3y After going through all the above, I think this is ok. tl;dr; is the old "retry 30 times behavior" would only kick in in corner cases where there's no great reason to keep retrying like that.

I'd put "Approved" but you're observing behavior of it taking ~ 4 mins to mark a node as down so we should understand that.

**Review by bstansberry — APPROVED** [review]
Ken's investigation showed the timing was as expected.
